pkg/v1/internal/estargz/estargz.go

```diff
@@ -0,0 +1,50 @@
+// Copyright 2020 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package estargz
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+
+	"github.com/containerd/stargz-snapshotter/estargz"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+)
+
+// ReadCloser reads uncompressed tarball input from the io.ReadCloser and
+// returns:
+//  * An io.ReadCloser from which compressed data may be read, and
+//  * A v1.Hash with the hash of the estargz table of contents, or
+//  * An error if the estargz processing encountered a problem.
+//
+// Refer to estargz for the options:
+// https://pkg.go.dev/github.com/containerd/stargz-snapshotter@v0.2.0/estargz#Option
+func ReadCloser(r io.ReadCloser, opts ...estargz.Option) (io.ReadCloser, v1.Hash, error) {
+	defer r.Close()
+
+	bs, err := ioutil.ReadAll(r)
```

Comment:
I imagine that sometimes r here is an `os.File` that already implements `io.SectionReader`, can we do:

```go
if sr, ok := r.(io.SectionReader); ok {
  rc, err = estargz.Build(...)
}
```

And avoid buffering sometimes?

Reply:
I thought about this, but `io.SectionReader` is not an interface, so this does not work.

I also thought about checking `io.ReaderAt`, but we still need to know the length.

If `os.File` had a length method (closest is `Stat`), we could define our own interface that wrapped `io.ReaderAt` and included the length method, but that doesn't work terribly well either 😞

Reply:
https://github.com/google/go-containerregistry/issues/876